util/src/linear_map.rs

```diff
@@ -0,0 +1,54 @@
+use alloc::vec::Vec;
+use core::mem;
+
+/// O(n) Vec-backed map for keys that only implement Eq.
+/// Only use this for a very small number of keys.
+pub struct LinearMap<K, V>(Vec<(K, V)>);
+
+impl<K, V> Default for LinearMap<K, V> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<K: Eq, V> LinearMap<K, V> {
+    pub fn new() -> Self {
+        Default::default()
+    }
+    pub fn get(&self, k: &K) -> Option<&V> {
+        self.0.iter().find(|(kk, _)| kk == k).map(|(_, v)| v)
+    }
+    pub fn get_mut(&mut self, k: &K) -> Option<&mut V> {
+        self.0.iter_mut().find(|(kk, _)| kk == k).map(|(_, v)| v)
+    }
+    pub fn insert(&mut self, k: K, mut v: V) -> Option<V> {
+        if let Some(vv) = self.get_mut(&k) {
+            mem::swap(&mut v, vv);
+            Some(v)
+        } else {
+            self.0.push((k, v));
+            None
+        }
+    }
+    pub fn values(&self) -> impl Iterator<Item = &V> {
+        self.0.iter().map(|(_, v)| v)
+    }
+}
+
+impl<K: Eq, V> FromIterator<(K, V)> for LinearMap<K, V> {
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
```

Comment:
Might be best to collect into a `Vec`, to avoid the scans `insert` would do. Not sure if performance matters here, but if it's about as simple

Reply:
Yeah, but then you can end up with duplicate keys. I can add more comments warning that it's n^2

Reply:
Oh right